internal/transport/http2_client.go

```diff
@@ -556,6 +556,13 @@ func (t *http2Client) createHeaderFields(ctx context.Context, callHdr *CallHdr)
 	// Make the slice of certain predictable size to reduce allocations made by append.
 	hfLen := 7 // :method, :scheme, :path, :authority, content-type, user-agent, te
 	hfLen += len(authData) + len(callAuthData)
+	registeredCompressors := t.registeredCompressors
+	if registeredCompressors != "" {
+		hfLen++
+	}
+	if _, ok := ctx.Deadline(); ok {
+		hfLen++
+	}
```

Comment:
Does it make sense to also check for `callHdr.SendCompress != ""` and `callHdr.PreviousAttempts > 0` and increment as well?

Reply:
Added these checks as well. 

---
I only added the `grpc-timeout` header initially since it was the one showing up in the profiles while running the benchmark. This was probably because compression was disabled and retries can't be configured in the benchmark. The improvements were similar after adding checks for the two conditions mentioned above and running the benchmark again with compression enabled.